Enable nxp low power comparator driver

ZhaoxiangJin opened the pull request:
The comparator now has its own driver API. We are planning to remove the sensor-based `mcux_lpcmp `driver and introduce the comparator-based `comparator_nxp_lpcmp `driver. In this PR, we:

1. Removed the `mcux_lpcmp `sample.
2. Marked the `mcux_lpcmp `driver as **deprecated** (will be deprecated in Zephyr 4.6).
3. Implemented the `comparator_nxp_lpcmp `driver.
4. Migrated all boards/SoCs that previously supported the `mcux_lpcmp `driver and sample to support the `comparator_nxp_lpcmp `driver and `gpio_loopback `test.

ZhaoxiangJin committed:
drivers: clock_control: Enable clock control for lpcmp

1. Enable MCXA/MCXN platforms' LPCMP clock control through
the clock driver (syscon).
2. Enable MCXE platform's LPCMP clock control through the
clock driver (mc_cgm).

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: comparator: enable nxp lpcmp driver

enable nxp lpcmp driver based on the comparator driver API

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

bjarki-andreasen reviewed (changes requested):
While I like the sensor based driver being removed in favor of the comparator API, Vestas has also contributed to the driver, maybe others want to maintain it? @henrikbrixandersen

ZhaoxiangJin committed:
doc: release-notes-4.4: Record mcux_lpcmp driver change

Update the release-notes-4.4 to document:

1. The mcux_lpcmp driver will be deprecated.
2. The compatible 'nxp,lpcmp' for mcux_lpcmp
has now been changed to 'nxp,sensor-lpcmp'.
3. A new lpcmp driver based on the comparator
driver API has been created, and the 'nxp,lpcmp'
compatible is used for it.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
tests: comparator: enable gpio_loopback for nxp boards

enable gpio_loopback for nxp boards, including:
- frdm_mcxa153
- frdm_mcxa156
- frdm_mcxa266
- frdm_mcxa346
- frdm_mcxa366
- frdm_mcxe31b
- frdm_mcxn236
- frdm_mcxn947/mcxn947/cpu0/qspi
- frdm_mcxn947/mcxn947/cpu0
- mcx_n9xx_evk/mcxn947/cpu0/qspi
- mcx_n9xx_evk/mcxn947/cpu0

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
dts: bindings: rename nxp,lpcmp.yaml to nxp,sensor-lpcmp.yaml

We are planning to implement a new LPCMP driver based on the
comparator API and deprecate the current sensor-based driver.
We would like to use the nxp,lpcmp binding for the new
comparator-based driver implementation. To avoid naming conflicts,
we are renaming the current sensor binding to nxp,sensor-lpcmp.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: sensor: mark mcux_lpcmp as deprecated

We are planning to implement a new lpcmp driver based on the
comparator API and deprecate the current sensor-based driver.
We now mark the mcux_lpcmp as deprecated.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
dts: nxp: Update lpcmp nodes for some SoCs

1. We shall set the proper clock source for lpcmp.
2. The property '#io-channel-cells' is currently
not needed for LPCMP.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: nxp: remove the lpcmp nodes from the board level dts

We shall enable lpcmp in application level. Enabling lpcmp
at the board level doesn't make any sense, since the lpcmp
input is determined by the application.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
> > While I like the sensor based driver being removed in favor of the comparator API, Vestas has also contributed to the driver, maybe others want to maintain it? @henrikbrixandersen
> 
> Driver removal needs to go through a deprecation process.

Thanks for your information, I have updated the PR.
I have selected the `DEPRECATED `Kconfig option for `mcux_lpcmp `and added an entry in the Zephyr 4.4 release notes documenting this deprecation (official deprecation will occur in Zephyr 4.6).
Please help to review, thanks.

hakehuang commented:
frdm_mcxa346 failed

```
(.venv) ubuntu@zephyr-test-11:~/nxp/frdm_mcxa346$ cat /home/ubuntu/nxp/frdm_mcxa346/temp/frdm_mcxa346_twister_test_only/twister-out/frdm_mcxa346_mcxa346/zephyr/tests/drivers/comparator/gpio_loopback/drivers.comparator.gpio_loopback.nxp_lpcmp/handler.log
*** Booting Zephyr OS build v4.3.0-2350-gbbff45f1c4cf ***
Running TESTSUITE smf_tests
================================================*** Booting Zephyr OS build v4.3.0-2704-ga43a45a073bb ***
Running TESTSUITE comparator_gpio_loopback
===================================================================
START - test_get_output

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/comparator/gpio_loopback/src/test.c:51: comparator_gpio_loopback_test_get_output: comparator_get_output(test_dev) not equal to 0

 FAIL - test_get_output in 0.021 seconds
===================================================================
START - test_no_trigger_no_pending
 PASS - test_no_trigger_no_pending in 0.003 seconds
===================================================================
START - test_trigger_both_edges_pending

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/comparator/gpio_loopback/src/test.c:103: comparator_gpio_loopback_test_trigger_both_edges_pending: comparator_trigger_is_pending(test_dev) not equal to 0

 FAIL - test_trigger_both_edges_pending in 0.020 seconds
===================================================================
START - test_trigger_callback

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/comparator/gpio_loopback/src/test.c:124: comparator_gpio_loopback_test_trigger_callback: k_sem_take(&test_sem, K_NO_WAIT) is non-zero

 FAIL - test_trigger_callback in 0.021 seconds
===================================================================
START - test_trigger_falling_edge_pending

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/comparator/gpio_loopback/src/test.c:82: comparator_gpio_loopback_test_trigger_falling_edge_pending: comparator_trigger_is_pending(test_dev) not equal to 0

 FAIL - test_trigger_falling_edge_pending in 0.020 seconds
===================================================================
START - test_trigger_rising_edge_pending

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/comparator/gpio_loopback/src/test.c:68: comparator_gpio_loopback_test_trigger_rising_edge_pending: comparator_trigger_is_pending(test_dev) not equal to 0

 FAIL - test_trigger_rising_edge_pending in 0.020 seconds
===================================================================
TESTSUITE comparator_gpio_loopback failed.

------ TESTSUITE SUMMARY START ------

SUITE FAIL -  16.67% [comparator_gpio_loopback]: pass = 1, fail = 5, skip = 0, total = 6 duration = 0.105 seconds
 - FAIL - [comparator_gpio_loopback.test_get_output] duration = 0.021 seconds
 - PASS - [comparator_gpio_loopback.test_no_trigger_no_pending] duration = 0.003 seconds
 - FAIL - [comparator_gpio_loopback.test_trigger_both_edges_pending] duration = 0.020 seconds
 - FAIL - [comparator_gpio_loopback.test_trigger_callback] duration = 0.021 seconds
 - FAIL - [comparator_gpio_loopback.test_trigger_falling_edge_pending] duration = 0.020 seconds
 - FAIL - [comparator_gpio_loopback.test_trigger_rising_edge_pending] duration = 0.020 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
RunID: 2555ccab48e4f8e57697c6f2c788650a
PROJECT EXECUTION FAILED

```

hakehuang reviewed (dismissed):
board test pass on all support platforms

bjarki-andreasen reviewed (dismissed):
looks good, regarding the removed sample @kartben do we have rules for removing a sample which demos a driver which has just been deprecated? should the sample be deprecated or can it just be removed?

henrikbrixandersen reviewed (dismissed):
I think the sample should stay for now. We can keep by setting the Kconfig for allowing testing of deprecated features.

ZhaoxiangJin committed:
samples: sensor: mcux_lpcmp: Update board overlays for lpcmp

Update the board overlay files for mcux_lpcmp sample to use
the sensor-lpcmp driver:

1. Delete the 'compatible' and 'clocks' properties from
   the SoC-level lpcmp nodes.
2. Set the compatible to "nxp,sensor-lpcmp" to use the
   sensor-based lpcmp driver.
3. Add '#io-channel-cells = <2>' property for sensor API.
4. Add pinctrl configuration for lpcmp0.
5. Ensure status is set to "okay".

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

bjarki-andreasen reviewed (approved):
Nice work!

ZhaoxiangJin commented:
> How come my review was dismissed?


Because your reason for requesting the change was that you did not want the sample to be removed, and I also chose to listen to your opinion without seeing a clear process regulation, I dismissed it. If there were any other change requests without a conclusion at that time, I would not have chosen to dismiss them.

henrikbrixandersen commented on drivers/sensor/nxp/mcux_lpcmp/Kconfig:
```diff
@@ -4,15 +4,22 @@
+	  It is currently scheduled to be removed in Zephyr 4.6.
+	  We will implement a new LPCMP driver based on the comparator driver
+	  API and enable test cases for boards/SoCs that previously supported
+	  the mcux_lpcmp driver and sample.
```
This text does not really belong here.

henrikbrixandersen commented on drivers/sensor/nxp/mcux_lpcmp/Kconfig:
```diff
@@ -4,15 +4,22 @@
+
 config MCUX_LPCMP_TRIGGER
-	bool "Trigger support"
+	bool "Trigger support [DEPRECATED]"
```
It's enough to mark the the top-level Kconfig as deprecated.

ZhaoxiangJin replied:
so, this is a problem？

henrikbrixandersen commented on drivers/sensor/nxp/mcux_lpcmp/mcux_lpcmp.c:
```diff
@@ -5,7 +5,7 @@
  */
 
-#define DT_DRV_COMPAT nxp_lpcmp
+#define DT_DRV_COMPAT nxp_sensor_lpcmp
```
This compatible should not need to change.

henrikbrixandersen reviewed (commented):
I think there are multiple issues with this PR not caught in the review (which was cut short due to my request for changes being dismissed).

https://github.com/zephyrproject-rtos/zephyr/pull/101469 tries to remedy the renamed compatible, but clearly there is no consensus on that either.

I suggest reverting this PR, have the author resubmit it - and let the review process run to an end.
